process/tls_parser.cpp

```diff
@@ -54,291 +49,478 @@ uint64_t quic_get_variable_length(uint8_t *start, uint64_t &offset)
    }
 } // quic_get_variable_length
 
-bool TLSParser::tls_is_grease_value(uint16_t val)
+bool TLSParser::is_grease_value(uint16_t val)
 {
-   if (val != 0 && !(val & ~(0xFAFA)) && ((0x00FF & val) == (val >> 8))) {
-      return true;
-   }
-   return false;
+    return val != 0 && !(val & ~(0xFAFA)) && ((0x00FF & val) == (val >> 8));
 }
 
-void TLSParser::tls_get_quic_user_agent(TLSData &data, char *buffer, size_t buffer_size)
+bool TLSParser::parse(const uint8_t* packet, uint32_t length)
 {
-   // compute end of quic_transport_parameters
-   const uint16_t quic_transport_params_len = ntohs(*(uint16_t *) data.start);
-   const uint8_t *quic_transport_params_end = data.start + quic_transport_params_len
-     + sizeof(quic_transport_params_len);
+    m_packet_data = packet;
+    m_packet_length = length;
+    clear_parsed_data();
+
+    if (!parse_tls_header()){
+        return false;
+    }
+    if (!parse_tls_handshake()){
+        return false;
+    }
+
+    if (m_handshake->type != TLS_HANDSHAKE_CLIENT_HELLO
+        && m_handshake->type != TLS_HANDSHAKE_SERVER_HELLO) {
+        return false;
+    }
+
+    if (!parse_session_id()) {
+        return false;
+    }
+
+    if (!parse_cipher_suites()){
+        return false;
+    }
+    if (!parse_compression_methods()){
+        return false;
+    }
+
+    if (m_handshake->type == TLS_HANDSHAKE_CLIENT_HELLO){
+        return parse_client_hello_extensions();
+    }
+    return parse_server_hello_extensions();
+}
 
-   if (quic_transport_params_end > data.end) {
-      return;
-   }
+bool TLSParser::parse_tls_header() const noexcept
+{
+    const auto* tls_header = (const TLSHeader*) m_packet_data;
+    if (sizeof(TLSHeader) > m_packet_length || tls_header == nullptr || tls_header->type != TLS_HANDSHAKE || tls_header->version.major != 3 || tls_header->version.minor > 3) {
+        return false;
+    }
+    return true;
+}
 
-   uint64_t offset = 0;
-   uint64_t param  = 0;
-   uint64_t length = 0;
-
-   while (data.start + offset < quic_transport_params_end) {
-      param  = quic_get_variable_length((uint8_t *) data.start, offset);
-      length = quic_get_variable_length((uint8_t *) data.start, offset);
-      if (param == TLS_EXT_GOOGLE_USER_AGENT) {
-         if (length + (size_t) 1 > buffer_size) {
-            length = buffer_size - 1;
-         }
-         memcpy(buffer, data.start + offset, length);
-         buffer[length] = 0;
-         data.obejcts_parsed++;
-      }
-      offset += length;
-   }
-   return;
+bool TLSParser::parse_tls_handshake() noexcept
+{
+    auto* handshake = (const TLSHandshake*)(m_packet_data + sizeof(TLSHeader));
+
+    if ( sizeof(TLSHeader) + sizeof(TLSHandshake) > m_packet_length ||
+        !(handshake->type == TLS_HANDSHAKE_CLIENT_HELLO || handshake->type == TLS_HANDSHAKE_SERVER_HELLO) ||
+        handshake->version.major != 3 ||
+        handshake->version.minor < 1 ||
+        handshake->version.minor > 3) {
+        return false;
+    }
```

Comment:
Tady ty podminky mi prijdou celkem extrem.

Co to treba rozdelit do neceho takove:
```
if (hasHandshakeType(...) {
...
}

if (hasRequiredVersion()) {
...
}